drivers/adc/adc_mchp_xec.c

```diff
@@ -0,0 +1,334 @@
+		return -EINVAL;
+	}
+
+	adc_regs->sar_control_reg.seldiff = channel_cfg->differential;
+
+	return 0;
+}
+
+static bool adc_xec_validate_buffer_size(const struct adc_sequence *sequence)
+{
+	int chan_count = 0;
+	size_t buff_need;
+	u32_t chan_mask;
+
+	for (chan_mask = 0x80; chan_mask != 0; chan_mask >>= 1) {
+		if (chan_mask & sequence->channels) {
+			chan_count++;
+		}
+	}
+
+	buff_need = chan_count * sizeof(u16_t);
+
+	if (sequence->options) {
+		buff_need *= 1 + sequence->options->extra_samplings;
+	}
+
+	if (buff_need > sequence->buffer_size) {
+		return false;
+	}
+
+	return true;
+}
+
+static int adc_xec_start_read(struct device *dev,
+			      const struct adc_sequence *sequence)
+{
+	struct adc_xec_regs *adc_regs = ADC_XEC_REG_BASE(dev);
+	struct adc_xec_data *data = dev->driver_data;
+
+	if (sequence->channels & ~XEC_ADC_CHANNELS_MASK) {
+		return -EINVAL;
+	}
+
+	if (sequence->channels == 0UL) {
+		return -EINVAL;
+	}
+
+	if (!adc_xec_validate_buffer_size(sequence)) {
+		return -ENOMEM;
+	}
+
+	if (sequence->resolution == 12) {
+		adc_regs->sar_control_reg.sel_res = XEC_ADC_12_BITS;
+	} else if (sequence->resolution == 10) {
+		adc_regs->sar_control_reg.sel_res = XEC_ADC_10_BITS;
+	} else {
+		return -EINVAL;
+	}
+
+	data->channels = sequence->channels;
+	data->buffer = sequence->buffer;
+	data->repeat_buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_xec_read(struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	struct adc_xec_data *data = dev->driver_data;
+	int error;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	error = adc_xec_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+
+#if defined(CONFIG_ADC_ASYNC)
+static int adc_xec_read_async(struct device *dev,
+			      const struct adc_sequence *sequence,
+			      struct k_poll_signal *async)
+{
+	struct adc_xec_data *data = dev->driver_data;
+	int error;
+
+	adc_context_lock(&data->ctx, true, async);
+	error = adc_xec_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+#endif /* CONFIG_ADC_ASYNC */
+
+static void xec_adc_get_sample(struct device *dev)
+{
+	struct adc_xec_regs *adc_regs = ADC_XEC_REG_BASE(dev);
+	struct adc_xec_data *data = dev->driver_data;
+	u32_t idx;
+	u32_t channels = adc_regs->status_reg.ch_status;
+	u32_t ch_status = channels;
+	u32_t bit = find_lsb_set(channels);
+
+	while (bit != 0) {
+		idx = bit - 1;
+
+		*data->buffer = (u16_t)adc_regs->channel_read_reg[idx];
+		data->buffer++;
+
+		channels &= ~BIT(idx);
+		bit = find_lsb_set(channels);
+	}
+
+	/* Clear the status register */
+	adc_regs->status_reg.ch_status = ch_status;
+}
+
+static void adc_xec_isr(struct device *dev)
+{
+	struct adc_xec_regs *adc_regs = ADC_XEC_REG_BASE(dev);
+	struct adc_xec_data *data = dev->driver_data;
+
+	xec_adc_get_sample(dev);
+
+	adc_regs->control_reg.single_done_status = 1;
```

Comment:
Clearing single done status with a bit-field has the side effect of clearing all other write-1-to-clear bits in this register.

Reply:
Changed to not use bit-fields.